catboost/libs/train_lib/train_model.cpp

```diff
@@ -1152,14 +1150,24 @@ void TrainModel(
         fstrType = EFstrType::PredictionValuesChange;
     }
 
+    THolder<NPar::TLocalExecutor> localExecutorHolder;
+    const ETaskType taskType = NCatboostOptions::GetTaskType(updatedTrainJson);
+    const bool isGpuDeviceType = taskType == ETaskType::GPU;
+    if(isGpuDeviceType && TTrainerFactory::Has(ETaskType::GPU)) {
+        localExecutorHolder.Reset(new NPar::TLocalExecutor());
+        localExecutorHolder.Get()->RunAdditionalThreads(catBoostOptions.SystemOptions.Get().NumThreads.Get() - 1);
+    } else {
+        localExecutorHolder.Reset(new NPar::TTbbLocalExecutor(catBoostOptions.SystemOptions.Get().NumThreads.Get() + 1));
+    }
+
```

Comment:
1. please put this code into `static THolder<NPar::ILocalExecutor> CreateLocalExecutor(const TCatBoostOptions& catboostOptions)` to avoid copy-paste
2. it is better to use `... = MakeHolder<TLocalExecutor>()` and `... = MakeHolder<TTbbLocalExecutor>(threadCount)` instead of `Reset`

Reply:
done